### PR TITLE
Parser support for render props

### PR DESCRIPTION
**Pull request by bkrmendy**
## Problem

In order to support editing operations on render props in the inspector, and making it possible to show render props in the navigator, we need to parse render props as JSX elements

## Fix
Parse them as JSX elements.

### Details
- The parser is changed so that if it detect a prop that's a JSX element, it parses it as such. The corresponding change is made in the printer too.
- the PR extends `JSExpression` with `JSXElement`. We'll need to support fragments and conditionals too in the future, this PR takes an incremental step toward that goal.
- Because of the change above, `jsxAttributeToValue` calls `renderCoreElement`. To make that possible, `jsxAttributeToValue` is passed all the arguments `renderCoreElement` needs
- The change above caused a circular dependency problem, so some functions have been shuffled around to work around that. `text-editor-utils.tsx` and `jsx-attribute-utils.tsx` were created to break to circular dependency chains

### CODE REVIEW

File: editor/src/core/model/element-template-utils.ts [hunk 1]
```diff
@@ -1275,7 +1284,12 @@ export function elementChildSupportsChildrenAlsoText(
   metadata: ElementInstanceMetadataMap,
   elementPathTree: ElementPathTrees,
 ): ElementSupportsChildren | null {
-  if (isJSExpression(element)) {
+  if (
+    isJSExpression(element) &&
+    // the condition below is a kludge to make the reparent tests pass, discuss
```
Comment (seanparsons) [inline editor/src/core/model/element-template-utils.ts:1289]: Still needs a discussion?

**Review by seanparsons — APPROVED** [review]

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap [hunk 2]
```diff
@@ -47297,10 +47297,7 @@ exports[`UiJsxCanvas render renders correctly when a component is passed in via
           style=\\"background-color: pink; width: 100%; height: 20px\\"
           data-path=\\"eee/fff/app-entity:aaa/bbb:ddd\\"
         >
-          <div
-            data-uid=\\"ccc~~~1\\"
-            data-path=\\"eee/fff/app-entity:aaa/bbb/ccc~~~1\\"
-          >
+          <div data-uid=\\"ccc\\" data-path=\\"eee/fff/app-entity:aaa/bbb/ccc\\">
```
Comment (balazsbajorics) [inline editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:47300]: is this change desired here?

File: editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-element-renderer-utils.tsx [hunk 3]
```diff
@@ -751,10 +762,12 @@ export function utopiaCanvasJSXLookup(
...
 
 function runJSExpression(
-  filePath: string,
-  requireResult: MapLike<any>,
   block: JSExpression,
+  elementPath: ElementPath | null,
   currentScope: MapLike<any>,
+  renderContext: RenderContext,
+  uid: string | undefined,
+  codeError: Error | null,
```
Comment (balazsbajorics) [inline editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-element-renderer-utils.tsx:770]: I would leave a comment here that this should probably deleted in a further refactor

File: editor/src/core/workers/parser-printer/parser-printer-parsing.ts [hunk 4]
```diff
@@ -2366,6 +2366,39 @@ function getAttributeExpression(
...
+      if (
+        parseResult.value.value.length !== 1 ||
+        parseResult.value.value[0].value.type !== 'JSX_ELEMENT'
+      ) {
+        return left('Cannot parse jsx element')
+      }
+
+      return right(
+        withParserMetadata(
+          parseResult.value.value[0].value,
```
Comment (balazsbajorics) [inline editor/src/core/workers/parser-printer/parser-printer-parsing.ts:2396]: shouldn't we assign `parseResult.value.value` to a named const for readability?

File: editor/src/core/workers/parser-printer/uid-fix.ts [hunk 5]
```diff
@@ -71,6 +71,8 @@ const expressionFunctionCallUIDOptic: Optic<JSExpressionFunctionCall, string> =
 const expressionOtherJavaScriptUIDOptic: Optic<JSExpressionOtherJavaScript, string> =
   fromField('uid')
 
+const expressionJSXElementOptic: Optic<JSXElement, string> = fromField('uid')
```
Comment (balazsbajorics) [inline editor/src/core/workers/parser-printer/uid-fix.ts:74]: I think this is missing a "uid" from the name 
```suggestion
const expressionJSXElementUIDOptic: Optic<JSXElement, string> = fromField('uid')
```

File: editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap (hunk 2 above)
Comment (gbalint) [reply editor/src/components/canvas/__snapshots__/ui-jsx-canvas.spec.tsx.snap:47300]: Yes. Before this jsx render props were wrapped into an expression (which just returned the element), which made them a generated element.
After this change these elements will not be considered generated (which is a prerequisite to make them editable).

File: editor/src/utils/value-parser-utils.ts [hunk 6]
```diff
@@ -325,6 +325,15 @@ export function parseJsx(_: unknown, value: unknown): ParseResult<JSXParsedValue
         name: type,
       })
     }
+    if (value.hasOwnProperty('props')) {
```
Comment (balazsbajorics) [inline editor/src/utils/value-parser-utils.ts:328]: is this a somewhat unrelated change?

**Review by balazsbajorics — APPROVED** [review]
looks very nice :)

File: editor/src/core/workers/parser-printer/uid-fix.ts (hunk 5 above)
Comment (gbalint) [reply editor/src/core/workers/parser-printer/uid-fix.ts]: I removed it, it was unused.

File: editor/src/utils/value-parser-utils.ts (hunk 6 above)
Comment (gbalint) [reply editor/src/utils/value-parser-utils.ts:328]: Yes, this was just a small issue we found during the more extensive jsx prop testing.
However, after this PR we may rethink on only relying on the runtime value and not on the parsed model in `parseJsx`. But that can come in a subsequent PR.
